### DISCUSSION TITLE
Auto download a file on wifi#connected

### DISCUSSION THREAD

**Discussion by paolomeraviglia** [arendst/Tasmota#21325] [Q&A]
Hi there,
I have a sonoff th10 in a remote place without stable wifi connection and would like to automate some on/off functions using timers.
The state depends by electricity cost, 'cause the relay will control a control line (not the load) for my EV charging box that will let the charger to run or to shut.
The requirements are these:
1) the provider publish every midday the timers for the next day that would define high rate times and low rate. I should enable timers accordingly and update them in an automated way, pushing timers for the running day and the next one.
2) the device will not be connected to internet exept when I will arrive with my car at the garage, so my android phone will turn up automatically an hotspot providing connection as soon as my android auto will disconnect.

The process I thought about is this:
- the sonoff will find the wifi connection, and using a rule triggered by Wifi#connected will download a file that will contain some commands (example provided below)
- the commands will be interpreted accordingly (like with function UfsRun) and the file will be generated on runtime (so, the time will be out of sync just by seconds... fair enough)

Actual limits:
- UrlFetch seems to work just on ESP32, so cannot be used
- TH10 has a 1M Flash, so no UFS is enabled by default
 
What is working:
- I already have the server part that will produce the file containing all the commands

Is there a way to achieve this with a sonoff TH10? Ideas?
The file will be 1-2KB.
I though also about using `webgetconfig`, but seems a little more complicated also with the automatic reboot after the config download.

here the file example:
```
Timers 1
Timer1 {"Enable":1,"Mode":0,"Time":"00:00","Window":0,"Days":"-----1-","Repeat":0,"Output":1,"Action":1}
Timer2 {"Enable":1,"Mode":0,"Time":"05:00","Window":0,"Days":"-----1-","Repeat":0,"Output":1,"Action":0}
Timer3 {"Enable":1,"Mode":0,"Time":"22:00","Window":0,"Days":"-----1-","Repeat":0,"Output":1,"Action":1}
Timer4 {"Enable":1,"Mode":0,"Time":"03:00","Window":0,"Days":"------1","Repeat":0,"Output":1,"Action":0}
Timer5 {"Enable":1,"Mode":0,"Time":"18:00","Window":0,"Days":"------1","Repeat":0,"Output":1,"Action":1}
Timer6 {"Enable":0 }
Timer7 {"Enable":0 }
Timer8 {"Enable":0 }
Timer9 {"Enable":0 }
Timer10 {"Enable":0 }
Timer11 {"Enable":0 }
Timer12 {"Enable":0 }
Timer13 {"Enable":0 }
Timer14 {"Enable":0 }
Timer15 {"Enable":0 }
Timer16 {"Enable":0 }
Power1 0
Time <epoch>
Backlog Latitude 47.4133024; Longitude 8.656394; TimeDST 0,0,3,1,1,120; TimeSTD 0,0,10,1,1,60; TimeZone 99
```

**Comment by lalo-uy** [comment]
Where ate you going to fetch the remote file?
Could be there a mqtt app suscribed to device online notification, and send
the mqtt timer cmds?

**Reply by paolomeraviglia** [reply]
Thanks for the answer,
an mqtt instance could be a possible solution, but would increase the complexity and maintenance over time.
For now, I could self host a static http server for this, but I was also thinking about pushing that file and update that file on a dedicated GitHub repo.
in the GitHub repo case, ignore the lines "Time <epoch>" and "Power1 0", that would not work as intended.

**Comment by lalo-uy** [comment]
You can also push mqtt retained messages, that will be quipped at 5he
server until the device logs in.

**Comment by barbudor** [comment]
There's no easy solution for that on ESP8266 for now
I'm having a look to bring an equivalent of Berry's UrlFetch() to ESP8266

**Comment by Jason2866** [comment]
Doing via mqtt is probably not a pain to maintain. Grab the file with a system "translate" and publish the info to the mqtt broker. The Tasmota device fetches the info and adjust the timers accordingly.

**Comment by barbudor** [comment]
@paolomeraviglia 
I have proposed a new command `WebRun` that would directly execute your command file.
See #21364

It still need to be approved

Note that it will not be included in any default build and you will have to custom-build to get the feature

**Reply by paolomeraviglia** [reply]
Great!
Thanks for the effort! I will test that asap!

**Reply by barbudor** [reply]
Beware it's not yet approved and merged
If you want, you can pull the code from my fork but if it isn't accepted you will end up with something stalled
